nntrainer/models/neuralnet.cpp

```diff
@@ -309,8 +322,13 @@ void NeuralNetwork::backwarding(sharedConstTensors input,
   forwarding(input, label);
 
   sharedConstTensors output = label;
-  for (unsigned int i = layers.size() - 1; i > 0; i--)
+  for (unsigned int i = layers.size() - 1; i > skip_non_trainable_layers + 1;
+       i--)
     output = layers[i]->backwarding(output, iteration);
+
+  /** The last trainable layer need not calculate the derivatives */
+  layers[skip_non_trainable_layers + 1]->calcGradient(output);
```

Comment:
Have you consider the api point of view? If neuralnet starts to know about this, we now have `layer::forward`, `layer::calcDeriv`, `layer::calcGradient` and `layer::applyGradient`. Also we have  `layer::backward` function but most likely it should be inherited from `nntrainer::internal`

Reply:
From the API point of view, `calDerivative`, `calcGradient` and `applyGradient` is exposed via layer_internal to neuralnet but for the API, it can be exposed or cannot be. It is still abstracted by `backwarding` and hidden under it from the external API point of view. `backwarding` function maintains is behavior.